### DISCUSSION TITLE
Generate every simple cycle from cycle basis in undirected graph

### DISCUSSION THREAD

**Discussion by chuymtz** [networkx/networkx#6394] [Algorithms]
Does networkx have a way to return every simple cycle from the cycle basis? The graph I have is not directed.

**Comment by dschult** [comment]
There are an infinite number of cycles. And every cycle can be written as a linear combination of the cycle basis elements a1*c1 + a2*c2+a3*c3+ ...+an*cn.  where ai*ci means you go around cycle ci in a clockwise direction ai times.  Negative ai means go around counterclockwise. There is no requirement that a1c1+a2c2 be a single connected cycle however -- e.g. they might be in different connected components. So, you'd need to rule those out if you don't consider those as cycles. Since you are restricting yourself to simple cycles and direction doesn't matter, you are restricting yourself to ai=1 or 0.

So you'll need to build the cycles from the basis yourself. NetworkX does not have such a function (and you should think hard about what exactly you want because there are many possible criteria for what counts as a simple cycle).  I'd start with a generator that loops over the subsets of the basis elements. Form the resulting "cycle" and then determine whether it is a kind you are interested in.

```python
def all_simple_cycles(basis, G):
    M = len(basis)
    for k in range(M):
        for cycles in itertools.combinations(basis, k):
            combined_cycle = []
            for cycle in cycles:
                combined_cycle.extend(cycle)
            # determine whether combined_cycle is what you want
            yield combined_cycle
```
This sounds pretty difficult to me. But hopefully it is easy to specify what type of cycle you want.
My advice is to think hard about what you really want.

**Comment by chuymtz** [comment]
Thank you for your answer. I believe I understand the issue but I'm not sure how to implement my specific criteria. This is a an example of what I want. 
With the following snippet I create an example with an undesired output

```
nodes = [0, 1, 2, 3, 4, 5, 6, 7]
edges = [[0,2],[1,2],[0,3],[1,3],[0,4],[4,5],[5,6],[1,7],[6,7],[5,1]]
G = nx.Graph()
G.add_nodes_from(nodes)
G.add_edges_from(edges)
nx.draw(G, with_labels=True)
```

Essentially what I want is the set of polygons in the network. That is 
```
[[2, 1, 5, 4, 0], [3, 1, 2,, 0], [6, 7, 1, 5]]
```
I run something similar to your suggestion but when it comes to a cycle such as 
```
 [3, 1, 5, 4, 0]
```
I'm not sure how to discard it given that there exists edges `[1,2]` and `[2,0]` inside the above. 
I want every polygon in the network and perhaps I'm not formulating the problem correctly. 

Thanks!

**Comment by dschult** [comment]
Good -- it is nice to have a concrete example -- and the fact that it is small is very helpful too.

The polygon [3, 1, 2, 0] is constructed via [3, 1, 5, 4, 0] - [2, 1, 5, 4, 0].  The subtraction removes the edges (1, 5) and (5, 4) and (4, 0) from the leading cycle while subtracting (adding the opposite edges) the edges (2, 1) and (0, 2) to give the cycle [3, 1, 2, 0].

There are also other cycles in this network.  For example, [3, 1, 7, 6, 5, 4, 0] = [3, 1, 5, 4, 0] - [6, 7, 1, 5]. But there are many others.  

The reason I said you need to think hard about what you want is that you chose the set of polygons [[2, 1, 5, 4, 0], [3, 1, 2, 0], [6, 7, 1, 5]].  But there are so many others that there's no way for the algorithm to know which set of 3 cycles you want. Theorems from linear algebra ensure that any chosen cycle basis for this network will have 3 cycles.  But just as in 3-D you can choose axes x, y, z or you can choose your axes u=x+y, v=z-y, w=y.  The difference is a change of "coordinate system" for the cycles of the network. 

How did you choose [[2, 1, 5, 4, 0], [3, 1, 2, 0], [6, 7, 1, 5]].  And do you really need that set of 3? Or could another set of 3 work just as well? Is there anything special about [3, 1, 2, 0] that would give it preference over [3, 1, 5, 4, 0] as the third cycle to consider?

**Comment by chuymtz** [comment]
Thank you again for your answer. These I chose because when i draw this on top of the original network there is no edge "crossing" through it. I felt i needed to perform some sort of operation between the original graph and the candidate cycle and see if there's at least one edge going through. So i need a cycle that cannot be further partitioned. 

The example shown in this post is what I want to do using networkx

https://mathematica.stackexchange.com/questions/45112/finding-all-chordless-simple-cycles-in-a-graph

I hope I'm making sense. The application I have is regarding a network covering a region of a map. We want to extract every simple polygon within it that does not overlap with any other. 

In the example I mention above, the cycle ` [3, 1, 5, 4, 0]` in the basis set is not valid because in the original graph there exist edge [1,2] and [2, 4] creating a path which "intersects" ` [3, 1, 5, 4, 0]`

**Comment by dschult** [comment]
Remember that a drawing of a network is a 2D projection of what is N-dimensional...  From a strictly Graph Theory perspective, we could draw the network differently (position the nodes differently) and the edges (1,2) and (2,0) do not cross the cycle.  Graph theory doesn't specify anything about the position of the nodes in a 2-D plane. To get results that depend on the "2-D embedding of a graph into the plane" you'll need algorithms that take position of the nodes into account.  Our cycle finding algorithms don't restrict the node positions, so they won't be able to take "intersections of edges" into account.  (if i understand what you mean by intersection.)

**Reply by chuymtz** [reply]
yes you understand correctly. I wish I could append the drawing that resembles what I see on the map. But you got the idea. Basically at this point I don't understand how to implement this criteria checker.

**Comment by dschult** [comment]
I know that for the [four color theorem](https://en.wikipedia.org/wiki/Four_color_theorem) (which turns maps into graphs and then colors them), instead of tracking edges on the map and turning that into a graph, they track regions (countries) and make those the nodes. Since the map is planar, the resulting graph should be planar (without edges intersecting -- just meeting at nodes).

This might allow you to reformulate the problem in terms of the nodes of a country graph instead of cycles around the boundary edges of a map. Your mileage may vary....

**Comment by smichr** [comment]
I'm looking at this thinking that it is very similar to what I am trying to do. DEEP told me that there isn't a formal name for it though it is like trying to find staircase polygons on a grid. The way I approach your problem is to label all the edges and construct sets with the labels in them (so if edge [0,1] is labeled "a" then and "a" goes in sets 0 and 1. The way I labeled things these are the sets:
```
['cjb', 'aehd', 'cd', 'aj', 'ib', 'ihg', 'gf', 'fe']
```
For purpose of the next step it is better if you use integers instead of letters: with `{'a': 0, 'b': 1, 'c': 2, 'd': 3, 'e': 4, 'f': 5, 'g': 6, 'h': 7, 'i': 8, 'j': 9}` we get
```
[{9, 2, 1}, {0, 3, 4, 7}, {2, 3}, {0, 9}, {8, 1}, {8, 6, 7}, {5, 6}, {4, 5}]
```
Let the sets be numbered 0-7 (your nodes) and call those x and the elements in each set y. You can now plot those points in 2-D and the rectangular regions which start at one point and travel only horizontally or vertically and try to come back to the start. I get the following nodes as being involved in the sets of "loops".

(5, 1, 0, 3, 4)
(1, 2, 0, 3)
(7, 0, 1, 3, 4, 6, 5)
(1, 7, 0, 4, 6, 5, 2)
(1, 5, 0, 4, 2)
(1, 7, 6, 5)

It would take more work to find edges (or sequences) that split what are for me the sufficient subset for my purposes.

**Comment by smichr** [comment]
Here is a working solution for the example that you gave:

```python
import networkx as nx

def find_minimal_cycles(G):
    """
    Identifies the smallest set of cycles that covers all nodes in the graph.
    """
    all_cycles = nx.cycle_basis(G)  # Get all cycles
    covered_nodes = set()
    selected_cycles = []

    while len(covered_nodes) < len(G.nodes()):
        # Find the smallest uncovered cycle
        smallest_cycle = min(
            (cycle for cycle in all_cycles if not set(cycle).issubset(covered_nodes)),
            key=len,
            default=None
        )

        if smallest_cycle is None:
            break  # No more cycles to add

        selected_cycles.append(smallest_cycle)
        covered_nodes.update(smallest_cycle)

    return selected_cycles

# Example usage:
G = nx.Graph()
edges = [(0, 2), (0, 3), (0, 4), (0, 2), (1, 2), (1, 2), (1, 3), (1, 5), (1, 7),
         (0, 3), (1, 3), (0, 4), (4, 5), (1, 5), (4, 5), (5, 6), (5, 6), (6, 7),
         (1, 7), (6, 7)]
G.add_edges_from(set(edges))  # Remove duplicates

minimal_cycles = find_minimal_cycles(G)
print("Minimal cycles covering all nodes:", minimal_cycles) # [[2, 0, 3, 1], [6, 5, 1, 7], [5, 4, 0, 3, 1]]
```
These are not the exact 3 you are looking for but that is a consequence of issues with 2D representation as was discussed. But what we can do, however, is see that the sequence 0,3,1 -- a path from 0 to 1 -- could also be taken as 0-2-1. So an alternate form for 5,4,0,3,1 would be 5,4,0,2,1.
